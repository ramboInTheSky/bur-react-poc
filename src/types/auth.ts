import { User } from './user';

export interface Auth {
  username: string;
  password: string;
  profile?: User;
}
