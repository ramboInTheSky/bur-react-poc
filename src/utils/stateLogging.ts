export default (state: any, action: any) => {
    const env = process.env.NODE_ENV;
    if (env !== 'production' && env !== 'test') {
        console.groupCollapsed('%c Application state change', 'color: white; background: black; padding: 2px');
        console.log(`%cAction: ${action.type}`, 'color: purple');
        console.log('%cState: ', 'color: blue', state);
        console.groupEnd();
    }
}