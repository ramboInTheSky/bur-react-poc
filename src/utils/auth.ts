import { Auth } from '../types/auth';

const {
  getAccountInfo,
  initRegistration,
  login: sdkLogin,
  logout: sdkLogout,
  register: sdkRegister,
} = window.gigya.accounts;

export const login = ({ username, password }: Auth) => {
  return new Promise((resolve, reject) => {
    sdkLogin({
      loginID: username,
      password,
      callback: (response: any) => {
        if (response.errorCode === 0) {
          resolve(response.user);
        } else {
          reject('Something went wrong');
        }
      },
    });
  });
};

export const logout = () => {
  return new Promise((resolve, reject) => {
    sdkLogout({
      callback: (response: any) => {
        if (response.errorCode === 0) {
          resolve();
        } else {
          reject('Something went wrong');
        }
      },
    });
  });
};

export const register = ({ username, password, profile }: Auth) => {
  return new Promise((resolve, reject) => {
    initRegistration({
      callback: (response: any) => {
        if (response.errorCode === 0) {
          sdkRegister({
            email: username,
            password,
            profile,
            regToken: response.regToken,
            finalizeRegistration: true,
            format: 'jsonp',
            callback: (response: any) => {
              if (response.errorCode === 0) {
                resolve(response.user);
              } else {
                reject('Something went wrong');
              }
            },
          });
        } else {
          reject('Something went wrong');
        }
      },
    });
  });
};

export const getAccount = () => {
  return new Promise((resolve) => {
    getAccountInfo({
      callback: (response: any) => {
        if (response.errorCode === 0) {
          resolve(response.profile);
        } else {
          resolve(undefined);
        }
      },
    });
  });
}
