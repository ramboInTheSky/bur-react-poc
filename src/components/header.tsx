import React, { FC } from 'react';
import { Link } from "react-router-dom";
import styled from 'styled-components';

import { User } from '../types/user';
import { logout } from '../utils/auth';

const Container = styled.header`
  display: flex;
  justify-content: space-between;
`;
const Title = styled.h1``;

const Nav = styled.nav`
  ul {
    display: flex;
    list-style: none;

    li {
      padding: 10px;
    }
  }
`;

export const Logout = styled.button`
  background: none;
  border: 0;
  outline: none;
  color: red;
  cursor: pointer;
  :hover {
    color: blue;
  }
`;

interface Props {
  readonly title: string;
  readonly onLogout: () => void;
  readonly user?: User;
}

export const Header: FC<Props> = ({ onLogout, title, user }) => {
  const handleLogout = () => logout().then(() => onLogout());

  return (
    <Container>
      <Title>{title}</Title>
      <Nav>
        {!!user && (
          <p>
            welcome {user.email},{' '}
            <Logout onClick={handleLogout}>logout</Logout>
          </p>
        )}
        {!user && (
          <ul>
            <li><Link to="/">Login</Link></li>
            <li><Link to="/register">Register</Link></li>
          </ul>
        )}
      </Nav>
    </Container>
  );
};
