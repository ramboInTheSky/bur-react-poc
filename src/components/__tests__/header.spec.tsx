import React from 'react';
import { shallow } from 'enzyme';

import { Header, Logout } from '../header';

import * as auth from '../../utils/auth';
jest.mock('../../utils/auth', () => ({
    logout: jest.fn(() => Promise.resolve()),
}));

describe(Header, () => {
  const props = {
    title: 'hello',
    onLogout: jest.fn(),
  };

  it('should render correctly', () => {
    expect(shallow(<Header {...props} />)).toMatchSnapshot();
  });

  it('should render correctly with a user', () => {
    expect(shallow(<Header {...props} user={{ email: 'test@test.com' }} />)).toMatchSnapshot();
  });

  it('should perform logout correctly', (done) => {
    const comp = shallow(<Header {...props} user={{ email: 'test@test.com' }} />);
    const logoutComp = comp.find(Logout);
    logoutComp.simulate('click');
    expect(auth.logout).toHaveBeenCalled();
    setTimeout(() => {
        expect(props.onLogout).toHaveBeenCalled();
        done();
    });
  });
});
