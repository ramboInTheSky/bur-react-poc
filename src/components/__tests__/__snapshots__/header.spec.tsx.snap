// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render correctly 1`] = `
<styled.header>
  <styled.h1>
    hello
  </styled.h1>
  <styled.nav>
    <ul>
      <li>
        <Link
          to="/"
        >
          Login
        </Link>
      </li>
      <li>
        <Link
          to="/register"
        >
          Register
        </Link>
      </li>
    </ul>
  </styled.nav>
</styled.header>
`;

exports[`Header should render correctly with a user 1`] = `
<styled.header>
  <styled.h1>
    hello
  </styled.h1>
  <styled.nav>
    <p>
      welcome 
      test@test.com
      ,
       
      <styled.button
        onClick={[Function]}
      >
        logout
      </styled.button>
    </p>
  </styled.nav>
</styled.header>
`;
